### PR TITLE
[REVIEW] Change ways to access `ptr` in `Buffer` 

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description
This PR:
With the introduction of `copy-on-write`(https://github.com/rapidsai/cudf/pull/11718), and `spilling`, there are different ways each `Buffer` implementation expects the code to behave when a `.ptr` is accessed, i.e., in case of COW, a `ptr` can be accessed for read or write purposes, in case of spilling the ptr can be accessed with a spill lock and the buffer will be spillable after the spill lock goes out of scope during execution. For these reasons we introduced `ptr`, `mutable_ptr` and `data_array_view`, `_data_array_view`, `mask_array_view`, and `_mask_array_view`. With so many ways to access buffer ptr and array views, this has become quite difficult for devs to know when to use what unless you are fully familiar with the implementation details. It will also lead us to a lot of special case handling for `Buffer`, `SpillableBuffer`, and `CopyonWriteBuffer`. For this reason, we have decided to simplify fetching the pointer with a `get_ptr(mode="read"/"write")` API, fetching data & mask array views will also become methods that accept `mode` like `data_array_view(mode="read"/"write")` & `mask_array_view(mode="read"/"write")`. It is the expectation that when the caller passed "read", they don't tamper with the buffer or the memory it is pointing to. In the case of "write", they are good to mutate the memory it is pointing to.

Note that even with `mode="read"/"write"` the caller should still, if appropriate, acquire a spill lock for the duration of the access. If this is not done, and the buffer is a `SpillableBuffer`, it will permanently be marked as unspillable.

- [x] Introduces `get_ptr()` to replace `ptr` property. 
- [x] Replaces `data_array_view` & `mask_array_view` methods with `data_array_view(mode=r/w)` & `mask_array_view(mode=r/w)`

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/_lib/copying.pyx
Comment (wence-) [reply python/cudf/cudf/_lib/copying.pyx:768]: Ah wait, no, in the `SpillableBuffer` case we don't want to mark these data as unspillable, no? Although I guess it doesn't matter because the `gpu_data` object doesn't persist past the end of this function (unless it's tracked somewhere globally?), so what we're doing here is transferring ownership of the pointer?

File: python/cudf/cudf/core/column/categorical.py
Comment (wence-) [inline python/cudf/cudf/core/column/categorical.py:959]: Can we default to safe but slow `"write"`? Or does `"write"` not differ from `read` for this case?

File: python/cudf/cudf/core/column/numerical.py
Comment (wence-) [inline python/cudf/cudf/core/column/numerical.py:583]: Do we need to `acquire_spill_lock` around this?

File: python/cudf/cudf/core/df_protocol.py
Comment (wence-) [inline python/cudf/cudf/core/df_protocol.py:92]: This is because the dataframe interchange protocol is a zero-copy one, so an external consumer is getting a view?

File: python/cudf/cudf/core/frame.py
Comment (wence-) [inline python/cudf/cudf/core/frame.py:1451]: This should acquire a spill lock, I think, but it doesn't matter because `outcol` is new and doesn't escape this function.

File: python/cudf/cudf/utils/queryutils.py
Comment (wence-) [inline python/cudf/cudf/utils/queryutils.py:223]: These columns are only read in `make_aggregate_nullmask`, no? If yes, do we also need to acquire a spill lock?

**Review by wence- — CHANGES_REQUESTED** [review]
Some small queries.

File: python/cudf/cudf/core/df_protocol.py
Comment (galipremsagar) [reply python/cudf/cudf/core/df_protocol.py:92]: I'm not quite familiar with df interchange protocol if it has to be a zero-copy one. I just made an equivalent switch of `.ptr` to `mode='write` assuming the buffer should be marked as exposed at this point since we are handing off to another library, my understanding here can be wrong too. @shwina to confirm.

File: python/cudf/cudf/core/frame.py
Comment (galipremsagar) [reply python/cudf/cudf/core/frame.py:1451]: That's correct

File: python/cudf/cudf/_lib/column.pyx
Comment (galipremsagar) [reply python/cudf/cudf/_lib/column.pyx:339]: Switched to `internal_write`

File: python/cudf/cudf/core/column/numerical.py
Comment (galipremsagar) [reply python/cudf/cudf/core/column/numerical.py:588]: Yup, added 👍

File: python/cudf/cudf/core/buffer/buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/buffer.py]: Yup, made the change 👍

File: python/cudf/cudf/core/buffer/buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/buffer.py]: Done 👍

File: python/cudf/cudf/utils/queryutils.py
Comment (galipremsagar) [reply python/cudf/cudf/utils/queryutils.py]: Yup, this can be a read-only with spill lock. Made the change 👍

**Comment by galipremsagar** [comment]
> Some small queries.

Addressed 👍  This PR should be ready for another round of reviews.

File: python/cudf/cudf/core/buffer/buffer.py
Comment (madsbk) [inline python/cudf/cudf/core/buffer/buffer.py:217]: short doc here, would be good

**Review by madsbk — CHANGES_REQUESTED** [review]
Overall looks good but I don't think we need the `"internal_write"` mode.

File: python/cudf/cudf/core/buffer/spillable_buffer.py
Comment (wence-) [reply python/cudf/cudf/core/buffer/spillable_buffer.py:311]: Is the idea that if no `spill_lock` exists that is saying "I want to expose this pointer for unlimited duration", whereas if there _is_ a `spill_lock` then I am only exposing things for the lifetime of the spill lock? If so, then I agree that `internal_write` is not needed, and it is up to the caller to acquire a spill lock if they want a scope lifetime exposure.

File: python/cudf/cudf/core/buffer/spillable_buffer.py
Comment (wence-) [inline python/cudf/cudf/core/buffer/spillable_buffer.py:310]: Irrespective of the discussion below, I am not sure that this can be correct. Since the `spill_lock` created here doesn't escape this function, the lifetime of the lock is not long enough to guarantee that the returned pointer is safe for use on the device.

File: python/cudf/cudf/core/buffer/spillable_buffer.py
Comment (madsbk) [reply python/cudf/cudf/core/buffer/spillable_buffer.py:311]: Yes

File: python/cudf/cudf/core/buffer/spillable_buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/spillable_buffer.py]: Removed this code

File: python/cudf/cudf/core/buffer/spillable_buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/spillable_buffer.py]: Got it, this approach sounds good to me 👍

File: python/cudf/cudf/core/buffer/buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/buffer.py:217]: Added docstring

File: python/cudf/cudf/core/buffer/buffer.py
Comment (wence-) [inline python/cudf/cudf/core/buffer/buffer.py:270]: How about this rewording?

> If "write", the data pointed to may be modified by the caller. If "read", the data pointed to must not be modified by the caller. Failure to fulfil this contract will cause incorrect behaviour.

File: python/cudf/cudf/core/buffer/buffer.py
Comment (wence-) [inline python/cudf/cudf/core/buffer/buffer.py:276]: Link to the subclasses if that is easy?

File: python/cudf/cudf/core/column/categorical.py
Comment (wence-) [reply python/cudf/cudf/core/column/categorical.py:959]: Still `"read"` here AFAICT?

File: python/cudf/cudf/core/column/column.py
Comment (wence-) [inline python/cudf/cudf/core/column/column.py:137]: If this is a spillable buffer the caller still has to hold a spill lock, right?

File: python/cudf/cudf/core/column/column.py
Comment (wence-) [inline python/cudf/cudf/core/column/column.py:127]: Suggestion: make explicit what this means.

> If the caller wishes to modify the data returned through this view, they must pass mode="write".

File: python/cudf/cudf/core/column/timedelta.py
Comment (wence-) [inline python/cudf/cudf/core/column/timedelta.py:128]:
```suggestion
    @acquire_spill_lock
    def to_arrow(self) -> pa.Array:
```

File: python/cudf/cudf/core/window/rolling.py
Comment (wence-) [inline python/cudf/cudf/core/window/rolling.py:491]: Need to acquire a spill lock here (for the duration of the `window_sizes_from_offset` call)

File: python/cudf/cudf/core/window/rolling.py
Comment (wence-) [inline python/cudf/cudf/core/window/rolling.py:536]: And a spill lock her.

File: python/cudf/cudf/utils/applyutils.py
Comment (wence-) [inline python/cudf/cudf/utils/applyutils.py:144]: Need to acquire a spill lock for this function.

File: python/strings_udf/strings_udf/_typing.py
Comment (wence-) [inline python/strings_udf/strings_udf/_typing.py:119]: This will expose any buffer passed in permanently unless someone is holding a spill lock, but I don't know who should be doing so? Presumably it is the caller of the strings_udf kernel. Where is that @brandon-b-miller ?

**Review by wence- — COMMENTED** [review]
Getting there, thanks! Some more minor things I think.

File: python/cudf/cudf/core/column/column.py
Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py:137]: Right

File: python/cudf/cudf/core/buffer/buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/buffer.py:270]: Done 👍

File: python/cudf/cudf/core/buffer/buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/buffer.py:276]: Linked 👍

File: python/cudf/cudf/core/column/categorical.py
Comment (galipremsagar) [reply python/cudf/cudf/core/column/categorical.py:959]: Sorry forgot to do so. Thanks for catching it. Should be fixed now.

File: python/cudf/cudf/core/column/column.py
Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py:127]: Done 👍

File: python/cudf/cudf/core/column/column.py
Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py:160]: Done

File: python/strings_udf/strings_udf/_typing.py
Comment (brandon-b-miller) [reply python/strings_udf/strings_udf/_typing.py:119]: two places:
https://github.com/rapidsai/cudf/blob/branch-23.02/python/strings_udf/strings_udf/tests/test_string_udfs.py#L71
https://github.com/rapidsai/cudf/blob/branch-23.02/python/cudf/cudf/core/indexed_frame.py#L2131

File: python/cudf/cudf/utils/applyutils.py
Comment (galipremsagar) [reply python/cudf/cudf/utils/applyutils.py:146]: Done

File: python/cudf/cudf/core/window/rolling.py
Comment (galipremsagar) [reply python/cudf/cudf/core/window/rolling.py]: Done 👍

File: python/cudf/cudf/core/series.py
Comment (galipremsagar) [reply python/cudf/cudf/core/series.py:4967]: Added spill locks 👍

**Review by brandon-b-miller — APPROVED** [review]
Approving `strings_udf` change.

**Comment by galipremsagar** [comment]
> Getting there, thanks! Some more minor things I think.

Thanks for catching the missing locks, @wence- ! Should be ready for review now.

**Review by wence- — APPROVED** [review]

**Comment by wence-** [comment]
I think I am happy now. Thanks very much for your perseverance!

File: python/cudf/cudf/tests/test_spilling.py
Comment (madsbk) [inline python/cudf/cudf/tests/test_spilling.py:122]:
```suggestion
    buf.mark_exposed()
```

File: python/cudf/cudf/tests/test_spilling.py
Comment (madsbk) [inline python/cudf/cudf/tests/test_spilling.py:212]:
```suggestion
    buf.mark_exposed()  # Expose pointer and trigger unspill
```

File: python/cudf/cudf/core/buffer/buffer.py
Comment (madsbk) [inline python/cudf/cudf/core/buffer/buffer.py:257]: I think it is great that we are setting `mode` by keyword everywhere, let's make it mandatory?
```suggestion
    def get_ptr(self, *, mode="write") -> int:
```

File: python/cudf/cudf/core/buffer/spillable_buffer.py
Comment (madsbk) [inline python/cudf/cudf/core/buffer/spillable_buffer.py:272]:
```suggestion
    def get_ptr(self, *, mode="write") -> int:
```

File: python/cudf/cudf/core/column/column.py
Comment (madsbk) [inline python/cudf/cudf/core/column/column.py:417]:
```suggestion
        return self.mask_array_view(mode="read")
```

File: python/cudf/cudf/core/column/column.py
Comment (madsbk) [reply python/cudf/cudf/core/column/column.py:417]: No, I think it can be `mode="read"`.
But notice, the access mode doesn't affect the spillability anymore. It is only whether `get_ptr(...)` is called within `acquire_spill_lock` or not that determines if the buffer is marked unspillable temporarily or permanently.

File: python/cudf/cudf/tests/test_spilling.py
Comment (madsbk) [inline python/cudf/cudf/tests/test_spilling.py:591]:
```suggestion
    buffers[0].mark_exposed()
```

File: python/cudf/cudf/tests/test_spilling.py
Comment (madsbk) [inline python/cudf/cudf/tests/test_spilling.py:620]:
```suggestion
        buffers[i].mark_exposed()
```

**Review by madsbk — COMMENTED** [review]
Some minor suggestions

**Review by vyasr — COMMENTED** [review]
Mostly looks good. I've got a few questions and suggestions, but not blocking merging. I'll defer to the other reviewers for a final approval since they've been much more active on this PR.

File: python/cudf/cudf/core/df_protocol.py
Comment (shwina) [reply python/cudf/cudf/core/df_protocol.py:92]: There's no promise that the interchange protocol is zero-copy. In-fact, in the case of cross-device (CPU<->GPU) interchange, it cannot be.

That being said, I think any public API via which we expose the raw pointer to a buffer should mark the buffer exposed.

Comment (galipremsagar) [reply python/cudf/cudf/core/df_protocol.py:92]: Thanks for clarifying @shwina! This will mark the buffer as exposed.

File: python/cudf/cudf/core/column/column.py
Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py]: Switched to mandatory keyword-only arg like `get_ptr`

File: python/cudf/cudf/core/column/column.py
Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py]: Switched to mandatory keyword-only arg like `get_ptr`

File: python/cudf/cudf/core/buffer/buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/buffer.py]: Removed `get_` and renamed to `_readonly_proxy_cai_obj`

File: python/cudf/cudf/core/buffer/buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/buffer.py:257]: Maintaining a default was to be on the safe side when someone doesn't specify anything but yeah by making it keyword-only we have eliminated that problem. I don't have a strong preference here, happy to change it. @wence- WDYT? Do you have any preference here?

File: python/cudf/cudf/core/buffer/buffer.py
Comment (wence-) [reply python/cudf/cudf/core/buffer/buffer.py]: "fulfil" is a valid spelling (otherwise codespell would have complained!)

File: python/cudf/cudf/core/column/column.py
Comment (wence-) [inline python/cudf/cudf/core/column/column.py:150]:
```suggestion
        self, *, mode="write"
```

This one has still snuck through :)

Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py]: Yikes, Sorry!

File: python/cudf/cudf/core/buffer/buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/buffer.py:257]: It is currently a no-op, but then COW is enabled. What do we want the default behavior to do? i.e., when someone does `.get_ptr()` call?

**Comment by harrism** [comment]
@galipremsagar Can you add some detail on motivation and purpose in the PR description?  I came here because of the cuSpatial PR to adapt to this change, but can't find an explanation.

**Comment by galipremsagar** [comment]
> @galipremsagar Can you add some detail on motivation and purpose in the PR description? I came here because of the cuSpatial PR to adapt to this change, but can't find an explanation.

Hi Mark, Thanks for asking for these details. I have updated the PR description now to also contain the motivation behind these changes.

File: python/cudf/cudf/core/buffer/buffer.py
Comment (wence-) [reply python/cudf/cudf/core/buffer/buffer.py:257]: @vyasr is suggesting this signature:

```python
def get_ptr(self, *, mode) -> int:
    ...
```

This a required keyword-only argument. So it would be invalid to call `buffer.get_ptr()`: the caller must _always_ provide a `mode` argument, but can _only_ do so as a keyword argument.

I think this would be a fine thing to do.

Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/buffer.py]: Got it, thanks for clarifying. Dropped the default 👍

File: python/cudf/cudf/core/buffer/buffer.py
Comment (shwina) [inline python/cudf/cudf/core/buffer/buffer.py:244]: Honestly, I don't think we should add extra property here given that it's used only in one other function (do we expect more uses for it?)

If not, I think we can just use `cuda_array_interface_wrapper` in those places instead.

File: python/cudf/cudf/core/buffer/buffer.py
Comment (shwina) [inline python/cudf/cudf/core/buffer/buffer.py:274]:
```suggestion
```

I think it's better not to make any promises that `mode=` will always be ignored for `Buffer`.

File: python/cudf/cudf/core/buffer/buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/buffer.py:244]: It is only being used in `data_array_view` & `mask_array_view`, and one more place for copy-on-write scenarios: https://github.com/rapidsai/cudf/pull/12556.

File: python/cudf/cudf/core/buffer/buffer.py
Comment (galipremsagar) [reply python/cudf/cudf/core/buffer/buffer.py]: okay by me 👍

File: python/strings_udf/strings_udf/tests/test_string_udfs.py
Comment (shwina) [inline python/strings_udf/strings_udf/tests/test_string_udfs.py:73]: I think we should avoid setting this precedent; tests really shouldn't have to worry about spilling.

Comment (galipremsagar) [inline python/strings_udf/strings_udf/tests/test_string_udfs.py:73]:
```suggestion
    kernel.forall(len(data))(str_views, output)
```
Dropping adding of spill lock here for what @shwina brought up and consistency with cudf pytests too.

File: python/strings_udf/strings_udf/tests/test_string_udfs.py
Comment (galipremsagar) [inline python/strings_udf/strings_udf/tests/test_string_udfs.py:1]:
```suggestion
# Copyright (c) 2022, NVIDIA CORPORATION.
```

File: python/strings_udf/strings_udf/tests/test_string_udfs.py
Comment (galipremsagar) [inline python/strings_udf/strings_udf/tests/test_string_udfs.py:13]:
```suggestion
```

File: python/cudf/cudf/core/buffer/buffer.py
Comment (shwina) [reply python/cudf/cudf/core/buffer/buffer.py:244]: OK -- let's leave as is now, and we can take stock in the merge of `copy-on-write`.

**Review by shwina — APPROVED** [review]

File: python/strings_udf/strings_udf/tests/test_string_udfs.py
Comment (galipremsagar) [reply python/strings_udf/strings_udf/tests/test_string_udfs.py]: Discussed offline and removed this lock to keep things consistent where we are explicitly not testing for spilling ability.

**Pull request merged**
